pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1342,13 +1421,45 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
 	}
 	klog.V(4).Infof("provisionClaim[%s]: started", claimToClaimKey(claim))
 	opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-	startTime := time.Now()
-	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
-		return err
-	})
+	plugin, storageClass, err := ctrl.findProvisionablePlugin(claim)
+	if err != nil {
+		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, err.Error())
+		klog.Errorf("error finding provisioning plugin for claim %s: %v", claimToClaimKey(claim), err)
+		// failed to find the requested provisioning plugin, directly return err for now.
+		// controller will retry the provisioning in every syncUnboundClaim() call
+		// retain the original behavior of returning nil from provisionClaim call
+		return nil
+	}
+	// for CIS provisioning, latency metric will be reported after final binding happened
+	if plugin == nil || plugin.IsMigratedToCSI() {
+		// TEST case
+		ctrl.scheduleOperation(opName, func() error {
+			// create an opTimestamps without assigning provisionerName to get the current timestamp
+			opTimestamps := newOperationTimestamp("")
+			provisionerName, err := ctrl.provisionClaimOperationCSI(claim, plugin, storageClass)
+			opTimestamps.provisionerName = provisionerName
+			// adding an entry into operationTimestamps for metrics reporting if there exists no
+			// such record for current claim regardless whether the operation returned error or not
+			// add an entry to mark the provisioning request start time
+			claimKey := claimToClaimKey(claim)
+			provisionStartTsKey := ctrl.createOperationTimestampCacheKey("provision", claimKey)
+			if _, exists := ctrl.operationTimestamps.Get(provisionStartTsKey); !exists {
+				ctrl.operationTimestamps.Add(provisionStartTsKey, opTimestamps)
+			}
+			return err
+		})
+	} else /* plugin != nil && !plugin.IsMigratedToCSI() */ {
+		// record provision start time for operation metric reporting
+		startTime := time.Now()
```

Comment:
I want to be able to provide both total e2e time for both in-tree and CSI volumes, so I think it's better that we have a brand new metric that clearly states its intended purpose, and doesn't break consumers of the existing metric.  If we find that after introducing the new metric, if the old metric doesn't serve much purpose anymore, then we can deprecate and remove the old metric.

If someone was developing their CSI driver and used these metrics, they should already realize there's a problem when their metrics give almost 0 latency.

Reply:
SGTM. how does the name "vol_provision_total_seconds" sound? happy to take any other good name

Reply:
The current one is called: storage_operation_duration_seconds

What if we call the new one: storage_operation_total_duration_seconds

@gnufied wdyt?

Reply:
We can keep same metric name but change operation_name. So for example - the operation name for these new metrics could be `total_provision`. At least that is what I ended up doing in - https://github.com/kubernetes/kubernetes/pull/75384 . This ensures that new metrics appear as new time series and don't conflict with existing metrics.

Reply:
I would prefer a different metric name because the semantics of what we're measuring are different.

Users could be consuming the existing metric and aggregating across all operations.

Reply:
Yeah,  I see pros and cons. On one hand - introducing new metric names hinders discover-ability of metrics. What is the use of a metric if no one knows about it? 

Do you see aggregation of existing operation metrics as a meaningful data point?  Sums of those numbers don't make sense. Mean makes sense to some extent but then again, attach operation could dominate the mean. 

I do not have very strong opinions on this. I would prefer not to introduce new metric name but if existing metrics are being aggregated then surely we will have to introduce a new name.  In which case - "storage_operation_total_duration_seconds" works.

Reply:
We can add a release note and update documentation on the metrics we are adding.

Aggregation is useful for creating high level SLOs, like "all control plane operations"

Reply:
I was thinking about using an other operator "label" instead of creating a new metric, but it will mess up with existing aggregation as @msau42 mentioned. I agree the new metric reporting logic is of a different semantic, though concerned about visibility of the metric. Creating new metric will increase the complexity of in-tree implementation as well. Within this PR, ONLY for CSI workflow, the controller will record the metric, another enhancement needs to implement the in-tree logic. If no other concerns, I will go ahead and create a new metric name: "storage_operation_total_duration_seconds". @msau42 @gnufied please confirm